### DISCUSSION TITLE
Ipsec tunnel won't finish connecting.

### DISCUSSION THREAD

**Discussion by eons93** [strongswan/strongswan#1057] [Q&A]
First, I'm terribly sorry, I've tried learning as much as I could about ipsec and strongswan before making a post and wasting peoples time with this post as I'm afraid my problem is something really simple. For some context, I'm a student in school, working on my capstone and have used wireguard lots and  have a fairish understanding of linux/ubuntu but we have been assigned in our capstone to create an ipsec tunnel for the first time. 

I have two ubuntu server 20.04 routers setup in our classroom network. One is our DMZ router with other servers behind it and has public ip addresses on its internet facing side. The other router is my LAN router which has workstations behind it. I'm using this guide to make my tunnels, [https://www.linode.com/docs/guides/strongswan-vpn-server-install/](url) and it is giving me much better results then a different guide that was a couple years old. That said, when I run ``` ipsec up lan-client ``` I get this output: 
```
initiating IKE_SA lan-client[2] to 168.156.192.203
generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
sending packet: from 172.16.2.53[500] to 168.156.192.203[500] (1128 bytes)
received packet: from 168.156.192.203[500] to 172.16.2.53[500] (38 bytes)
	parsed IKE_SA_INIT response 0 [ N(INVAL_KE) ]
	peer didn't accept DH group ECP_256, it requested CURVE_25519
initiating IKE_SA lan-client[2] to 168.156.192.203
generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
sending packet: from 172.16.2.53[500] to 168.156.192.203[500] (1096 bytes)
received packet: from 168.156.192.203[500] to 172.16.2.53[500] (236 bytes)
parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
selected proposal: IKE:CHACHA20_POLY1305/PRF_HMAC_SHA2_512/CURVE_25519
local host is behind NAT, sending keep alives
remote host is behind NAT
establishing CHILD_SA lan-client{2}
generating IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr CPRQ(ADDR DNS) SA TSi TSr N(MOBIKE_SUP) N(ADD_4_ADDR) N(ADD_4_ADDR) N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
sending packet: from 172.16.2.53[4500] to 168.156.192.203[4500] (311 bytes)
received packet: from 168.156.192.203[4500] to 172.16.2.53[4500] (1248 bytes)
parsed IKE_AUTH response 1 [ EF(1/2) ]
received fragment #1 of 2, waiting for complete IKE message
received packet: from 168.156.192.203[4500] to 172.16.2.53[4500] (713 bytes)
parsed IKE_AUTH response 1 [ EF(2/2) ]
received fragment #2 of 2, reassembled fragmented IKE message (1896 bytes)
parsed IKE_AUTH response 1 [ IDr CERT AUTH EAP/REQ/ID ]
received end entity cert "CN=dmz-lan"
  using certificate "CN=dmz-lan"
no issuer certificate found for "CN=dmz-lan"
  issuer is "CN=dmz-lan"
no trusted RSA public key found for '168.156.192.203'
generating INFORMATIONAL request 2 [ N(AUTH_FAILED) ]
sending packet: from 172.16.2.53[4500] to 168.156.192.203[4500] (65 bytes)
establishing connection 'lan-client' failed
```

and this is what I get when I run ```ipsec up dmz-server``` on the dmz router (which probably means nothing because this end is supposed to be the responder, it can't find a client thats behind nat but whenever I try to find logs on either machine, i get nothing):
```
unable to resolve %any, initiate aborted
tried to checkin and delete nonexisting IKE_SA
establishing connection 'dmz-server' failed
```

DMZ 's .conf:
```
config setup
        charondebug="ike 1, knl 1, cfg 0, net 1"
        strictcrlpolicy=no
        uniqueids=yes
        cachecrls=no
conn dmz-server
      auto=add
      compress=no
      type=tunnel
      keyexchange=ikev2
      fragmentation=yes
      forceencaps=yes
      dpdaction=clear
      dpddelay=300s
      rekey=no
      left=%any
      leftid=[public ip address]
      leftcert=server.cert.pem
      leftsendcert=always
      leftsubnet=10.0.3.0/24
      right=%any
      rightid=%any
      rightsubnet=10.1.3.0/24
      rightauth=eap-mschapv2
      rightsourceip=172.16.2.53
      rightsendcert=never
      eap_identity=%identity
      ike=chacha20poly1305-sha512-curve25519-prfsha512,aes256gcm16-sha384-prfsha384-ecp384,aes256-sha1-modp1024,aes128-sha1-modp1024,3des-sha1-modp1024!
      esp=chacha20poly1305-sha512,aes256gcm16-ecp384,aes256-sha256,aes256-sha1,3des-sha1!
```

DMZ's .secrets
```
: RSA "/etc/ipsec.d/private/server.key.pem"
kylecc : EAP "pleaseWORK!"
```

LAN's .conf
```
config setup
conn lan-client
    right=[public ip address]
    rightid=[public ip address]
    rightsubnet=10.0.3.0/24
    rightauth=pubkey
    leftsourceip=%config
    leftid=kylecc
    leftauth=eap-mschapv2
    eap_identity=%identity
    auto=start
```

LAN's .secrets
```
kylecc : "pleaseWORK!"
```

again, apoplogies if I'm wasting peoples time with a really easy standard problem but I'm just so overwhelmed trying to figure this out, I could really use the help of it veterans to help me out. Hope everyone is having a blessed day.

**Comment by tobiasbrunner** [comment]
> ````
> received end entity cert "CN=dmz-lan"
>   using certificate "CN=dmz-lan"
> no issuer certificate found for "CN=dmz-lan"
>   issuer is "CN=dmz-lan"
>````

Is this server certificate self-signed? If it is, you have to install it locally in `/etc/ipsec.d/certs` and configure it in `rightcert` of the `lan-client` connection. If it's actually signed a CA certificate, you have to install that in `/etc/ipsec.d/cacerts` (note it's generally a good idea to use different Distinguished Names for CA and end-entity certificates).

You might want to have a look at [docs.strongswan.org](https://docs.strongswan.org) (e.g. the introduction and the quickstart howtos).

**Reply by eons93** [reply]
Yeah. the guide i was using and is linked above was much more helpful then others but had a few mistakes in it, like telling me to put the client side cert directly in ipsec.d directory, not in the cacerts directory. I do have a new problem though. Not entirely sure if I should just end this discussion and remake a new one but currently the tunnel is established, installed and both the 10.0.3.0/24 subnet and 10.1.3.0/24 subnet got that '===' thing between them when I run ```ipsec status``` but when I ping for a device across the tunnel, I just get ```destination host unreachable```. My research and what not is telling me I need to add some routes but that doesnt make sense to me and also not exactly sure what those routes should look like. Wouldn't the strongswan program have some kind of sensor to detect when traffic is intended for across the tunnel and route it as needed?

**Reply by tobiasbrunner** [reply]
strongSwan should install routes in table 220 automatically (see `ip route list table 220`). Depending on where that ICMP error is sent from there could be different issues (e.g. forwarding the traffic on either side). I'd recommend [this article](https://docs.strongswan.org/docs/5.9/howtos/forwarding.html) in the docs.
